src/coreclr/src/jit/emit.cpp

```diff
@@ -5691,39 +5697,93 @@ void emitter::emitDataGenEnd()
 #endif
 }
 
-/********************************************************************************
- * Generates a data section constant
- *
- * Parameters:
- *     cnsAddr  - memory location containing constant value
- *     cnsSize  - size of constant in bytes
- *     cnsAlign - alignment of constant in bytes
- *
- * Returns constant number as offset into data section.
- */
-UNATIVE_OFFSET emitter::emitDataConst(const void* cnsAddr, UNATIVE_OFFSET cnsSize, UNATIVE_OFFSET cnsAlign)
+//---------------------------------------------------------------------------
+// emitDataGenFind:
+//   - Returns the offset of an existing constant in the data section
+//     or INVALID_UNATIVE_OFFSET if there was no matching constant
+//
+// Arguments:
+//    cnsAddr    - A pointer to the value of the constant that we need
+//    cnsSize    - The size in bytes of the constant
+//    alignment  - The requested alignment for the data
+//    dataType   - The type of the constant int/float/etc
+//
+UNATIVE_OFFSET emitter::emitDataGenFind(const void* cnsAddr, unsigned cnsSize, unsigned alignment, var_types dataType)
 {
-    UNATIVE_OFFSET cnum = emitDataGenBeg(cnsSize, cnsAlign);
-    emitDataGenData(0, cnsAddr, cnsSize);
-    emitDataGenEnd();
+    UNATIVE_OFFSET cnum     = INVALID_UNATIVE_OFFSET;
+    unsigned       cmpCount = 0;
+    unsigned       curOffs  = 0;
+    dataSection*   secDesc  = emitConsDsc.dsdList;
+    while (secDesc != nullptr)
+    {
+        // Search the existing secDesc entries
+
+        if ((secDesc->dsSize == cnsSize) && (secDesc->dsDataType == dataType) && ((curOffs % alignment) == 0))
```

Comment:
Can we not just match the raw integral bits rather than needing to check "dataType"?

For example, if you had a `ulong` constant that was `0x8000_0000_0000_0000` you could also use it for a `double` constant that was `-0.0` as they have the same 64-bit pattern. This would likely be useful for scenarios where you are doing floating-point algorithms that are manipulating the underlying bits.

Likewise, I imagine this might be less efficient for certain SIMD code. If you emit a SIMD constant for `<0.0f, 1.0f, 2.0f, 3.0f>` and also have a floating-point constant for `1.0f` then we will have both emitted separately, rather than just having the latter load from element 1 of the SIMD constant.

These might just be future optimization opportunities, but I wanted to call them out.

Reply:
I could make that change.  It would probably match floating point zero to integer zero frequently.
I also could do `(secDesc->dsSize >= cnsSize) `as well.

Reply:
> It would probably match floating point zero to integer zero frequently

I think this is probably a rare case. On ARM there is the explicit zero register and on x86 I think we prefer `xor reg, reg` and `xorps reg, reg` as they are both smaller code and elided by the instruction decoder.